mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/upstream/processor/PublishProcessor.java

```diff
@@ -51,23 +57,49 @@ public class PublishProcessor implements UpstreamProcessor {
     @Resource
     private FirstTopicManager firstTopicManager;
 
+    @Resource
+    RetainedPersistManager retainedPersistManager;
 
     @Override
-    public CompletableFuture<HookResult> process(MqttMessageUpContext context, MqttMessage mqttMessage) {
+    public CompletableFuture<HookResult> process(MqttMessageUpContext context, MqttMessage mqttMessage) throws RemotingException, com.alipay.sofa.jraft.error.RemotingException, ExecutionException, InterruptedException {
         MqttPublishMessage mqttPublishMessage = (MqttPublishMessage) mqttMessage;
+        boolean isEmpty = false;
+        //deal empty payload
+        if (ByteBufUtil.getBytes(mqttPublishMessage.content()).length == 0) {
+            mqttPublishMessage = MessageUtil.dealEmptyMessage(mqttPublishMessage);
+            isEmpty = true;
+        }
         String msgId = MessageClientIDSetter.createUniqID();
         MqttPublishVariableHeader variableHeader = mqttPublishMessage.variableHeader();
         String originTopic = variableHeader.topicName();
         String pubTopic = TopicUtils.normalizeTopic(originTopic);
         MqttTopic mqttTopic = TopicUtils.decode(pubTopic);
-        firstTopicManager.checkFirstTopicIfCreated(mqttTopic.getFirstTopic());
-        Set<String> queueNames = wildcardManager.matchQueueSetByMsgTopic(pubTopic, context.getNamespace());
+        firstTopicManager.checkFirstTopicIfCreated(mqttTopic.getFirstTopic());      // Check the firstTopic is existed
+        Set<String> queueNames = wildcardManager.matchQueueSetByMsgTopic(pubTopic, context.getNamespace()); //According to topic to find queue
+        long bornTime = System.currentTimeMillis();
+
+        MqttPublishMessage retainedMqttPublishMessage = mqttPublishMessage.copy();
+
+        if (mqttPublishMessage.fixedHeader().isRetain()) {
+            //Change the retained flag of message that will send MQ is 0
+            mqttPublishMessage = MessageUtil.removeRetainedFlag(mqttPublishMessage);
+            //Keep the retained flag of message that will store meta
+            Message metaMessage = MessageUtil.toMessage(retainedMqttPublishMessage);
+            metaMessage.setMsgId(msgId);
+            metaMessage.setBornTimestamp(bornTime);
+            metaMessage.setEmpty(isEmpty);
+            CompletableFuture<Boolean> booleanCompletableFuture = retainedPersistManager.storeRetainedMessage(TopicUtils.normalizeTopic(metaMessage.getOriginTopic()), metaMessage);
```

Comment:
if store retain message error, send message action will failed. it should be notify cs and user